site/en/blog/crx-scripting-api/index.md

```diff
@@ -0,0 +1,289 @@
+page.
+
+```js
+// Manifest V2 extension
+let userReq = await fetch('https://example.com/greetUser.js');
+let userBody = await userReq.text();
+let user = eval(userBody);
+
+chrome.browserAction.onClicked.addListener((tab) => {
+  chrome.tabs.executeScript({
+    // user.greetUser == 'alert("Hello, <GIVEN_NAME>!")'
+    code: user.greetUser,
+  });
+});
+```
+
+While Manifest V3 extensions can't use code that isn't bundled with the extension, our goal was to
+preserve some of the dynamism that arbitrary code blocks enabled for Manifest V2 extensions. The
+function and arguments approach makes it possible for Chrome Web Store reviewers, users, and other
+interested parties to more accurately assess the risks an extension poses while also allowing
+developers to modify an extension's runtime behavior based on user settings or application state.
+
+```js
+// Manifest V3 extension
+let userReq = await fetch('https://example.com/user-data.json');
+let user = await userReq.json();
+let givenName = user.givenName || '<GIVEN_NAME>';
+
+function greetUser(name) {
+  alert(`Hello, ${name}!`);
+}
+chrome.action.onClicked.addListener((tab) => {
+  chrome.tabs.executeScript(tab.id, {
+    func: greetUser,
+    args: [givenName],
+  });
+});
+```
+
+### Targeting frames
+
+We also wanted to improve how developers interact with frames in the revised API. The Manifest V2
+version of `executeScript` allowed developers to target either all frames in a tab or a specific
+frame in the tab.
+
+```js
+// Manifest V2 extension
+chrome.browserAction.onClicked.addListener((tab) => {
+  chrome.tabs.executeScript(tab.id, {
+    frameId: targetFrame1,
+    file: 'content-script.js',
+  });
+  chrome.tabs.executeScript(tab.id, {
+    frameId: targetFrame2,
+    file: 'content-script.js',
+  });
+});
+```
+
+In Manifest V3, we replaced the optional `frame` integer property in the options object with an
+optional `frames` array of integers; this allows developers to target multiple frames in a single
+API call. You can use use `chrome.webNavigation.getAllFrames` to get a list of all frames in a tab.
+
+```js
+// Manifest V3 extension
+chrome.action.onClicked.addListener((tab) => {
+  chrome.scripting.executeScript({
+    target: {
+      tabId: tab.id,
+      frames: [targetFrame1, targetFrame2],
+    },
+    files: ['content-script.js'],
+  });
+});
+```
+
+### Script injection results
+
+We've also improved the way we return script injection results in Manifest V3. A "result" is
+basically the final statement evaluated in a script. Think of it like the value returned when you
+execute a block of code in the Chrome DevTools console.
+
+In Manifest V2, `executeScript` and `insertCSS` would return an array of bare execution results.
+This is fine for a single injection point, but if you're injecting into all frames in a tab it can
+be difficult to track down which frame encountered the error.
```

Comment:
Or even which frame gave which result, which I think is the more germane example (since we don't [ yet ] return errors from individual frames).  Though I like your approach to it below.  Maybe a combination?  Can we return something "meaningful" from the page, rather than just "success"?  e.g. number of occurrences of a word on a page, etc.

Reply:
I can definitely tweak the language here to be more generic. I actually learned about this capability while writing this post; the reference to catching errors in the last line here was based on my original, incorrect assumption about how results worked. 

I was intentionally trying to avoid having the script return meaningful data because I didn't want to encourage using this as a general purpose message channel. This was largely based on my own experimentation and being surprised by certain types behaving in was I didn't anticipate (e.g. Error instances converted to null). I'd like to get your thoughts before changing this.

Reply:
Spoke about this in the morning and iterated on the content in subsequent commits.